### PR TITLE
Support for SNS Expiration Notifications

### PR DESCRIPTION
**Pull request by jtschladen**
This change adds support for expiration notifications sent via AWS SNS. Topics are configured on a per-notification basis. Owner/security notifications will still be sent via email.

Here's a screenshot of the notification configuration:
Screen Shot 2020-10-20 at 12 05 10 PM

### CODE REVIEW

File: lemur/plugins/lemur_aws/plugin.py [hunk 1]
```diff
@@ -406,3 +407,54 @@ def upload(self, name, body, private_key, chain, options, **kwargs):
...
+
+        topic_arn = f"arn:aws:sns:{self.get_option('region', options)}:" \
+                    f"{self.get_option('accountNumber', options)}:" \
+                    f"{self.get_option('topicName', options)}"
+
+        current_app.logger.debug(f"Publishing {notification_type} notification to topic {topic_arn}")
+        try:
+            sns.publish(topic_arn, message, notification_type, region_name=self.get_option("region", options))
+        except Exception:
+            current_app.logger.exception(f"Error publishing {notification_type} notification to topic {topic_arn}")
```
Comment (charhate) [inline lemur/plugins/lemur_aws/plugin.py:460]: How about capturing to sentry as well?
`sentry.captureException()`

File: lemur/notifications/messaging.py [hunk 2]
```diff
@@ -30,7 +30,7 @@
 
 def get_certificates(exclude=None):
     """
-    Finds all certificates that are eligible for notifications.
+    Finds all certificates that are eligible for expiration notifications.
```
Comment (charhate) [inline lemur/notifications/messaging.py:33]: One thought about the query in this method, we should [add filter](https://github.com/Netflix/lemur/blob/master/lemur/notifications/messaging.py#L44) for `Certificate.revoked == True` as well.

File: lemur/plugins/lemur_aws/plugin.py (hunk 1 above)
Comment (jtschladen) [reply lemur/plugins/lemur_aws/plugin.py:460]: You make a good point, but I actually think I need to remove this try/catch instead. `messaging.py` is the only caller of this `send` method, and it has its own try/catch surrounding it that already alerts on failure. I'll remove it here - good catch, thanks.

File: lemur/notifications/messaging.py (hunk 2 above)
Comment (jtschladen) [reply lemur/notifications/messaging.py:33]: Makes sense, no reason to send any notifications for revoked certs.

**Review by charhate — APPROVED** [review]

File: lemur/plugins/lemur_aws/plugin.py [hunk 3]
```diff
@@ -406,3 +407,51 @@ def upload(self, name, body, private_key, chain, options, **kwargs):
...
+        """
+        While we receive a `targets` parameter here, it is unused, as the SNS topic is pre-configured in the
+        plugin configuration, and can't reasonably be changed dynamically.
+        """
+
+        topic_arn = f"arn:aws:sns:{self.get_option('region', options)}:" \
+                    f"{self.get_option('accountNumber', options)}:" \
+                    f"{self.get_option('topicName', options)}"
+
+        current_app.logger.debug(f"Publishing {notification_type} notification to topic {topic_arn}")
```
Comment (hosseinsh) [inline lemur/plugins/lemur_aws/plugin.py:456]: I wonder if this shouldn't be info

File: lemur/plugins/lemur_aws/sns.py [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+    return "https://{hostname}/#/certificates/{name}".format(
+        hostname=current_app.config.get("LEMUR_HOSTNAME"), name=name
+    )
+
+
+def format_message(certificate, notification_type):
+    json_message = {
+        "notification_type": notification_type,
+        "certificate_name": certificate["name"],
+        "expires": arrow.get(certificate["validityEnd"]).format("dddd, MMMM D, YYYY"),
```
Comment (hosseinsh) [inline lemur/plugins/lemur_aws/sns.py:50]: Do you think a more machine-friendly format would be valuable?
like:
```
yyyy'-'MM'-'dd'T'HH':'mm':'ss | 2016-03-22T06:30:07
```

File: lemur/notifications/messaging.py [hunk 5]
```diff
@@ -87,29 +89,30 @@ def get_eligible_certificates(exclude=None):
...
-        "message": f"Sending expiration notification for to targets {targets}",
+        "message": f"Sending expiration notification for to recipients {recipients}",
         "notification_type": "expiration",
-        "certificate_targets": targets,
+        "certificate_targets": recipients,
     }
     status = FAILURE_METRIC_STATUS
     try:
-        notification.plugin.send(event_type, data, targets, notification.options)
+        current_app.logger.debug(log_data)
```
Comment (hosseinsh) [inline lemur/notifications/messaging.py:111]: should this be info?

**Review by hosseinsh — COMMENTED** [review]
an elaborate and well-tested PR. added some minor suggestions. 
It looks good to me. 
if I can find another 20min, I would take a more detailed look at `messaging.py`, otherwise, everything looks good to me.

File: lemur/plugins/lemur_aws/plugin.py (hunk 3 above)
Comment (jtschladen) [reply lemur/plugins/lemur_aws/plugin.py:456]: I could imagine this being a high volume log if there are many of these notifications configured, but it is also pretty useful to have logged. I'll go ahead and switch it (as well as the Slack equivalent) to info, and we can tune it back down later if needed.

File: lemur/plugins/lemur_aws/sns.py (hunk 4 above)
Comment (jtschladen) [reply lemur/plugins/lemur_aws/sns.py:50]: Sure - I copied the message from the Slack plugin, but that does sound like a more useful format.

File: lemur/notifications/messaging.py (hunk 5 above)
Comment (jtschladen) [reply lemur/notifications/messaging.py:111]: Same concern about volume (but even more than SNS) - I'm imagining there could be dozens/hundreds/maybe thousands of email notifications sent at once, which could be more logging than is actually useful. That said, if you don't think it's a valid concern, I can change it.

Comment (hosseinsh) [reply lemur/notifications/messaging.py]: I see your point. This might be way higher volume than the SNS part.

**Review by hosseinsh — APPROVED** [review]

**Pull request merged**